test/common/stats/stats_matcher_test.cc

```diff
@@ -0,0 +1,245 @@
+#include "envoy/config/metrics/v2/stats.pb.h"
+
+#include "common/stats/stats_matcher_impl.h"
+
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+using testing::IsFalse;
+using testing::IsTrue;
+
+namespace Envoy {
+namespace Stats {
+
+class StatsMatcherTest : public testing::Test {
+protected:
+  envoy::type::matcher::StringMatcher* inclusionList() {
+    return stats_config_.mutable_stats_matcher()->mutable_inclusion_list()->add_patterns();
+  }
+  envoy::type::matcher::StringMatcher* exclusionList() {
+    return stats_config_.mutable_stats_matcher()->mutable_exclusion_list()->add_patterns();
+  }
+  void rejectAll(const bool should_reject) {
+    stats_config_.mutable_stats_matcher()->set_reject_all(should_reject);
+  }
+
+  void initMatcher() { stats_matcher_impl_ = std::make_unique<StatsMatcherImpl>(stats_config_); }
+
+  envoy::config::metrics::v2::StatsConfig stats_config_;
+
+protected:
+  void expectAccepted(std::vector<std::string> expected_to_pass) {
+    for (const auto& stat_name : expected_to_pass) {
+      EXPECT_FALSE(stats_matcher_impl_->rejects(stat_name)) << "Accepted: " << stat_name;
+    }
+  }
+  void expectDenied(std::vector<std::string> expected_to_fail) {
+    for (const auto& stat_name : expected_to_fail) {
+      EXPECT_TRUE(stats_matcher_impl_->rejects(stat_name)) << "Rejected: " << stat_name;
+    }
+  }
+
+private:
+  std::unique_ptr<StatsMatcherImpl> stats_matcher_impl_;
+};
+
+TEST_F(StatsMatcherTest, CheckDefault) {
+  // With no set fields, everything should be allowed through.
+  initMatcher();
+  expectAccepted({"foo", "bar", "foo.bar", "foo.bar.baz", "foobarbaz"});
+}
+
+// Across-the-board matchers.
+
+TEST_F(StatsMatcherTest, CheckRejectAll) {
+  // With reject_all, nothing should be allowed through.
+  rejectAll(true);
+  initMatcher();
+  expectDenied({"foo", "bar", "foo.bar", "foo.bar.baz", "foobarbaz"});
+}
+
+TEST_F(StatsMatcherTest, CheckNotRejectAll) {
+  // With !reject_all, everything should be allowed through.
+  rejectAll(false);
+  initMatcher();
+  expectAccepted({"foo", "bar", "foo.bar", "foo.bar.baz", "foobarbaz"});
+}
+
+TEST_F(StatsMatcherTest, CheckIncludeAll) {
+  inclusionList()->set_regex(".*");
+  initMatcher();
+  expectAccepted({"foo", "bar", "foo.bar", "foo.bar.baz"});
+}
+
+TEST_F(StatsMatcherTest, CheckExcludeAll) {
+  exclusionList()->set_regex(".*");
+  initMatcher();
+  expectDenied({"foo", "bar", "foo.bar", "foo.bar.baz"});
+}
+
+// Single exact matchers.
+
+TEST_F(StatsMatcherTest, CheckIncludeExact) {
+  inclusionList()->set_regex("abc");
```

Comment:
Should this be set_exact, since this is supposed to be an exact matcher?

Reply:
Great catch, fixed in 6686c4218.